cpp/src/io/parquet/writer_impl.cu

```diff
@@ -886,34 +926,33 @@ gpu::parquet_column_device_view parquet_column_view::get_device_view(
   return desc;
 }
 
-void writer::impl::init_page_fragments(cudf::detail::hostdevice_2dvector<gpu::PageFragment>& frag,
-                                       device_span<gpu::parquet_column_device_view const> col_desc,
-                                       host_span<partition_info const> partitions,
-                                       device_span<int const> part_frag_offset,
-                                       uint32_t fragment_size)
+void writer::impl::init_row_group_fragments(
+  cudf::detail::hostdevice_2dvector<gpu::PageFragment>& frag,
+  device_span<gpu::parquet_column_device_view const> col_desc,
+  host_span<partition_info const> partitions,
+  device_span<int const> part_frag_offset,
+  uint32_t fragment_size)
 {
   auto d_partitions = cudf::detail::make_device_uvector_async(partitions, stream);
-  gpu::InitPageFragments(frag, col_desc, d_partitions, part_frag_offset, fragment_size, stream);
+  gpu::InitRowGroupFragments(frag, col_desc, d_partitions, part_frag_offset, fragment_size, stream);
   frag.device_to_host(stream, true);
 }
 
-void writer::impl::gather_fragment_statistics(
-  device_2dspan<statistics_chunk> frag_stats_chunk,
-  device_2dspan<gpu::PageFragment const> frag,
-  device_span<gpu::parquet_column_device_view const> col_desc,
-  uint32_t num_fragments)
+void writer::impl::calculate_page_fragments(device_span<gpu::PageFragment> frag,
+                                            host_span<size_type const> frag_sizes)
 {
-  auto num_columns = col_desc.size();
-  rmm::device_uvector<statistics_group> frag_stats_group(num_fragments * num_columns, stream);
-  auto frag_stats_group_2dview =
-    device_2dspan<statistics_group>(frag_stats_group.data(), num_columns, num_fragments);
-
-  gpu::InitFragmentStatistics(frag_stats_group_2dview, frag, col_desc, stream);
-  detail::calculate_group_statistics<detail::io_file_format::PARQUET>(frag_stats_chunk.data(),
-                                                                      frag_stats_group.data(),
-                                                                      num_fragments * num_columns,
-                                                                      stream,
-                                                                      int96_timestamps);
+  auto d_frag_sz = cudf::detail::make_device_uvector_async(frag_sizes, stream);
+  gpu::CalculatePageFragments(frag, d_frag_sz, stream);
```

Comment:
So this takes a span made from a hostdevice_vector, builds a device vector from it and runs calculations on it? Why are we building this device vector?

Reply:
frag_sizes is a `host_span` from a `std::vector`, not a `hostdevice_vector`.  Just need to push it to device for this one kernel.

Reply:
Right, I was looking at the first argument at the call site. My mistake.